src/model_signing/signing/signing.py

```diff
@@ -188,19 +199,46 @@ def manifest_from_payload(
 
         Returns:
             A manifest that can be converted back to the same payload.
+
+        Raises:
+            ValueError: If the payload cannot be deserialized to a manifest.
         """
+        obtained_predicate_type = payload["predicateType"]
+        if obtained_predicate_type != cls.predicate_type:
+            raise ValueError(
+                f"Predicate type mismatch, expected {cls.predicate_type}, "
+                f"got {obtained_predicate_type}"
+            )
+
         subjects = payload["subject"]
+        if len(subjects) != 1:
+            raise ValueError(f"Expected only one subject, got {subjects}")
+
+        model_name = subjects[0]["name"]
+        expected_digest = subjects[0]["digest"]["sha256"]
 
+        predicate = payload["predicate"]
+        serialization_dict = predicate["serialization"]
+        serialization = manifest.SerializationType.from_dict(serialization_dict)
+
+        hasher = memory.SHA256()
         items = []
-        for subject in subjects:
-            path = pathlib.PurePosixPath(subject["name"])
-            algorithm = subject["annotations"]["actual_hash_algorithm"]
-            digest_value = subject["digest"]["sha256"]
+        for resource in predicate["resources"]:
+            name = resource["name"]
+            algorithm = resource["algorithm"]
+            digest_value = resource["digest"]
             digest = hashing.Digest(algorithm, bytes.fromhex(digest_value))
-            item = manifest.FileManifestItem(path=path, digest=digest)
-            items.append(item)
+            hasher.update(digest.digest_value)
+            items.append(serialization.new_item(name, digest))
+
+        obtained_digest = hasher.compute().digest_hex
+        if obtained_digest != expected_digest:
```

Comment:
im guessing this is cheap to compute on just digest of digests, but do we need to verify it's consistent?

Reply:
We need to, otherwise someone could change the root digest to match the digest from another document that does not have the full manifest and attach fake properties to the model